providers/implementations/keymgmt/ml_kem_kmgmt.c

```diff
@@ -0,0 +1,523 @@
+    int genok = 0;
+
+    if (gctx == NULL
+        || (key = ml_kem_new(gctx->libctx, gctx->propq, gctx->variant)) == NULL)
+        return NULL;
+
+    /* Actual keypair generation may optionally be deferred */
+    if ((gctx->selection & OSSL_KEYMGMT_SELECT_KEYPAIR) == 0)
+        return key;
+
+    genok = seed != NULL ?
+        ossl_ml_kem_genkey_seed(seed, slen, nopub, 0, key) :
+        ossl_ml_kem_genkey_rand(seed, slen, nopub, 0, key);
+
+    /* Erase the single-use seed */
+    if (seed)
+        OPENSSL_cleanse(seed, ML_KEM_SEED_BYTES);
+    gctx->seed = NULL;
+
+    if (genok)
+        return key;
+
+    ossl_ml_kem_key_free(key);
+    return NULL;
+}
+
+static void ml_kem_gen_cleanup(void *vgctx)
+{
+    struct ml_kem_gen_ctx *gctx = vgctx;
+
+    if (gctx->seed != NULL)
+        OPENSSL_cleanse(gctx->seed, ML_KEM_RANDOM_BYTES);
+    OPENSSL_free(gctx->propq);
+    OPENSSL_free(gctx);
+}
+
+static void *ml_kem_dup(const void *vkey, int selection)
+{
+    const ML_KEM_KEY *key = vkey;
+
+    if (!ossl_prov_is_running())
+        return NULL;
+
+    return ossl_ml_kem_key_dup(key, selection);
+}
+
+typedef void (*fPtr)(void);
+
+#define DECLARE_VARIANT(bits) \
+    static void *ml_kem_##bits##_new(void *provctx) \
+    { \
+        return ml_kem_new(provctx == NULL ? NULL : PROV_LIBCTX_OF(provctx), \
+                          NULL, ML_KEM_##bits); \
+    } \
+    static void *ml_kem_##bits##_gen_init(void *provctx, int selection, \
+                                          const OSSL_PARAM params[]) \
+    { \
+        return ml_kem_gen_init(provctx, selection, params, ML_KEM_##bits); \
+    } \
+    const OSSL_DISPATCH ossl_ml_kem_##bits##_keymgmt_functions[] = { \
+        { OSSL_FUNC_KEYMGMT_NEW, (fPtr) ml_kem_##bits##_new }, \
+        { OSSL_FUNC_KEYMGMT_FREE, (fPtr) ossl_ml_kem_key_free }, \
+        { OSSL_FUNC_KEYMGMT_GET_PARAMS, (fPtr) ml_kem_get_params }, \
+        { OSSL_FUNC_KEYMGMT_GETTABLE_PARAMS, (fPtr) ml_kem_gettable_params }, \
+        { OSSL_FUNC_KEYMGMT_SET_PARAMS, (fPtr) ml_kem_set_params }, \
+        { OSSL_FUNC_KEYMGMT_SETTABLE_PARAMS, (fPtr) ml_kem_settable_params }, \
+        { OSSL_FUNC_KEYMGMT_HAS, (fPtr) ml_kem_has }, \
+        { OSSL_FUNC_KEYMGMT_MATCH, (fPtr) ml_kem_match }, \
+        { OSSL_FUNC_KEYMGMT_GEN_INIT, (fPtr) ml_kem_##bits##_gen_init }, \
+        { OSSL_FUNC_KEYMGMT_GEN_SET_PARAMS, (fPtr) ml_kem_gen_set_params }, \
+        { OSSL_FUNC_KEYMGMT_GEN_SETTABLE_PARAMS, \
+            (fPtr) ml_kem_gen_settable_params }, \
+        { OSSL_FUNC_KEYMGMT_GEN, (fPtr) ml_kem_gen }, \
+        { OSSL_FUNC_KEYMGMT_GEN_CLEANUP, (fPtr) ml_kem_gen_cleanup }, \
+        { OSSL_FUNC_KEYMGMT_DUP, (fPtr) ml_kem_dup }, \
+        { OSSL_FUNC_KEYMGMT_IMPORT, (fPtr) ml_kem_import }, \
+        { OSSL_FUNC_KEYMGMT_IMPORT_TYPES, (fPtr) ml_kem_imexport_types }, \
+        { OSSL_FUNC_KEYMGMT_EXPORT, (fPtr) ml_kem_export }, \
+        { OSSL_FUNC_KEYMGMT_EXPORT_TYPES, (fPtr) ml_kem_imexport_types }, \
```

Comment:
At some point we may need to check if OSSL_FUNC_KEYMGMT_VALIDATE needs to be implemented.

Reply:
I don't think it is possible to create an inconsistent keypair, but if the method needs to be there, it could always return 1.